File: android/src/main/java/expo/modules/heartrate/HeartRateModule.kt
package expo.modules.heartrate

import android.os.Build
import android.os.Handler
import android.os.Looper
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlin.random.Random

class HeartRateModule : Module() {
  private val wearManager by lazy { WearDataLayerManager(appContext.reactContext!!) }
  private var isMonitoring = false
  private var simulationHandler: Handler? = null
  private var simulationRunnable: Runnable? = null
  private var simulatedBPM = 72.0

  private val isEmulator: Boolean
    get() = Build.FINGERPRINT.contains("generic") ||
            Build.FINGERPRINT.contains("emulator") ||
            Build.MODEL.contains("Emulator") ||
            Build.MODEL.contains("sdk_gphone")

  override fun definition() = ModuleDefinition {
    Name("HeartRate")

    Events("heartRateUpdate", "connectionChange", "error")

    OnCreate {
      HeartRateEventBridge.register { event ->
        if (isMonitoring) {
          val zoneStatus = HeartRateZoneCalculator.getZoneStatus(event.bpm.toInt())
          sendEvent("heartRateUpdate", mapOf(
            "bpm" to event.bpm,
            "timestamp" to event.timestamp,
            "source" to "wearOS",
            "zone" to zoneStatus,
          ))
        }
      }
    }

    OnDestroy {
      HeartRateEventBridge.unregister()
      stopSimulation()
    }

    Function("startMonitoring") {
      isMonitoring = true
      if (isEmulator) {
        startSimulation()
      } else {
        wearManager.sendStartCommand()
      }
    }

    Function("stopMonitoring") {
      isMonitoring = false
      if (isEmulator) {
        stopSimulation()
      } else {
        wearManager.sendStopCommand()
      }
    }

    AsyncFunction("isWatchConnected") { promise: expo.modules.kotlin.Promise ->
      if (isEmulator) {
        promise.resolve(true)
      } else {
        wearManager.checkConnectivity { connected ->
          promise.resolve(connected)
        }
      }
    }

    AsyncFunction("getHeartRateZones") {
      HeartRateZoneCalculator.getZonesAsMaps()
    }
  }

  private fun startSimulation() {
    simulatedBPM = 72.0
    simulationHandler = Handler(Looper.getMainLooper())
    simulationRunnable = object : Runnable {
      override fun run() {
        if (!isMonitoring) return
        val delta = Random.nextDouble(-3.0, 5.0)
        simulatedBPM = (simulatedBPM + delta).coerceIn(55.0, 185.0)

        val zoneStatus = HeartRateZoneCalculator.getZoneStatus(simulatedBPM.toInt())
        sendEvent("heartRateUpdate", mapOf(
          "bpm" to simulatedBPM,
          "timestamp" to System.currentTimeMillis(),
          "source" to "wearOS",
          "zone" to zoneStatus,
        ))
        simulationHandler?.postDelayed(this, 1000)
      }
    }
    simulationHandler?.post(simulationRunnable!!)
  }

  private fun stopSimulation() {
    simulationRunnable?.let { simulationHandler?.removeCallbacks(it) }
    simulationHandler = null
    simulationRunnable = null
  }
}

File: ios/HeartRateModule.swift
import ExpoModulesCore

public class HeartRateModule: Module {
  private let watchManager = WatchConnectivityManager.shared
  private let zoneCalculator = HeartRateZoneCalculator.shared
  private var isMonitoring = false

  private var isSimulator: Bool {
    #if targetEnvironment(simulator)
    return true
    #else
    return false
    #endif
  }

  public func definition() -> ModuleDefinition {
    Name("HeartRate")

    Events("heartRateUpdate", "connectionChange", "error")

    OnCreate {
      self.watchManager.delegate = self
      self.watchManager.activate()
    }

    Function("startMonitoring") { (config: [String: String]?) in
      self.zoneCalculator.initialize { _ in }
      self.isMonitoring = true
      self.watchManager.sendStartCommand(config: config)
    }

    Function("stopMonitoring") {
      self.isMonitoring = false
      self.watchManager.sendStopCommand()
    }

    AsyncFunction("isWatchConnected") { () -> Bool in
      if self.isSimulator { return true }
      return self.watchManager.isWatchReachable
    }

    AsyncFunction("getHeartRateZones") { () -> [[String: Any]] in
      return self.zoneCalculator.getZones()
    }
  }
}

// MARK: - WatchConnectivityDelegate

extension HeartRateModule: WatchConnectivityDelegate {
  func didReceiveHeartRate(bpm: Double, timestamp: TimeInterval) {
    guard isMonitoring else { return }

    let zoneStatus = zoneCalculator.getZoneStatus(bpm: Int(bpm))

    sendEvent("heartRateUpdate", [
      "bpm": bpm,
      "timestamp": timestamp,
      "source": "watchOS",
      "zone": zoneStatus,
    ])
  }

  func didChangeReachability(isReachable: Bool) {
    sendEvent("connectionChange", [
      "isConnected": isReachable,
    ])
  }

  func didEncounterError(message: String, code: String) {
    sendEvent("error", [
      "message": message,
      "code": code,
    ])
  }
}
